pkg/apis/volumesnapshot/v1alpha1/types.go

```diff
@@ -0,0 +1,334 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshotList is a list of VolumeSnapshot objects
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of VolumeSnapshots
+	Items []VolumeSnapshot `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+	// PersistentVolumeClaimName is the name of the PVC being snapshotted
+	// +optional
+	PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotContent
+	// +optional
+	SnapshotDataName string `json:"snapshotDataName" protobuf:"bytes,2,opt,name=snapshotDataName"`
+
+	// Name of the SnapshotClass required by the volume snapshot.
+	// +optional
+	SnapshotClassName string `json:"snapshotClassName" protobuf:"bytes,3,opt,name=snapshotClassName"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SnapshotClass describes the parameters used by storage system when
+// provisioning VolumeSnapshots from PVCs.
+//
+// SnapshotClasses are non-namespaced; the name of the snapshot class
+// according to etcd is in ObjectMeta.Name.
+type SnapshotClass struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Snapshotter is the driver expected to handle this SnapshotClass.
+	Snapshotter string `json:"snapshotter" protobuf:"bytes,2,opt,name=snapshotter"`
+
+	// Parameters holds parameters for the snapshotter.
+	// These values are opaque to the system and are passed directly
+	// to the snapshotter.
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty" protobuf:"bytes,3,rep,name=parameters"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SnapshotClassList is a collection of snapshot classes.
+type SnapshotClassList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of SnapshotClasses
+	Items []SnapshotClass `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshotContent represents the actual "on-disk" snapshot object
+type VolumeSnapshotContent struct {
```

Comment:
I know we had some discussion on this in the last call. Wouldn't VolumeSnapshotClaim or VolumeSnapshotRequest as the user-facing object and VolumeSnapshot as the storage backend representation be better names? These names follow the PersistentVolumeClaim-PersistentVolume convention and can be pluralized.

Reply:
Unfortunately, PersistentVolumeClaim does not follow the name pattern as all other user-facing objects. For example, we don't use PodClaim to represent user's pod. We think we should have named PVC and PV as PersistentVolume (user) and PersistentVolumeContent (system admin) at the first place. So instead of following the wrong pattern, we decided to use VolumeSnapshot to represent the user's snapshot.

Reply:
PVCs don't follow other user-facing objects because other objects don't follow the dual-object model where we have a request and a response or a user-facing representation and an internal representation, but I agree with you that it's better not to follow the wrong pattern and have the more straightforward name (i.e., VolumeSnapshot) exposed to the users.

Reply:
OK I'm late to to conversation so don't really have much right to input. I would prefer sticking to the PVC/PV pattern even if you belive it is wrong as it exists now. I don't belive it is incorrect as @kangarlou says noting else follows the request response pattern. It will as